staging/src/k8s.io/apimachinery/pkg/api/validate/strfmt.go

```diff
@@ -65,3 +66,28 @@ func DNSSubdomain[T ~string](_ context.Context, op operation.Operation, fldPath
 	}
 	return allErrs
 }
+
+// GenerateName verifies that the specified value passes the validation
+// function when being treated as a Kubernetes generateName prefix.
+func GenerateName[T ~string](ctx context.Context, op operation.Operation, fldPath *field.Path, value, _ *T, fn ValidateFunc[*T]) field.ErrorList {
+	if value == nil {
+		return nil
+	}
+	modified := T(maskTrailingDash((string)(*value)))
+	return fn(ctx, op, fldPath, &modified, nil)
+}
+
+func maskTrailingDash(base string) string {
+	// This is dumb. If the final validation asserts max-length=63
+	// (DNS label), this will allow a 63 byte base, but not a 64 byte base,
+	// despite the fact that they are both going to be truncated to 58 (63-5)
+	// and then get a 5 character suffix.
+	//
+	// This bug has existed since forever, but we could ratchet-in the proper
+	// behavior, which would be to mimic the name generation logic and validate
+	// the result.
```

Comment:
+100

I'm a fan of carrying the cruft through the migration to declarative validation.  Then, after migrated, we can use declarative validation to ratchet-in the proper behavior.  This has the advantage of showcasing the benefits of declarative validation.